### ISSUE TITLE
SupportedLocales: returned Array is weird

### ISSUE THREAD

**Issue by jakobkummerow** [tc39/ecma402#258]
Steps 5 through 8 of SupportedLocales (https://tc39.github.io/ecma402/#sec-supportedlocales) create an Array from the List of locales that the function wants to return, then iterate over that Array to make all its own properties (i.e. elements and length) non-writable and non-configurable.

I guess this predates the introduction of `Object.freeze`, because notably it leaves the object extensible: while user code can't delete or modify any of the existing array elements, it can add additional properties to the array.

That said, why is it useful to attempt to write-protect the array in the first place? While there's probably no good reason for user code to want to modify the array returned by an `Intl.Something.supportedLocalesOf()` call, I also don't see a good reason to preemptively make this impossible. If a program wanted to change the result array, it could always create a copy first, and then modify that copy at will. So the operation could be simplified by dropping steps 6 and 7.

In summary, I present the following options for your consideration:

(1) Change nothing, keep the spec as is.
(2) Replace the custom loop around `DefinePropertyOrThrow` with a call to `Object.freeze`.
(3) Simply return a regular mutable array (as returned by CreateArrayFromList).

My personal preference would be (3), because simplification is nice.

Implementation complexity is not a concern; we already have the code for (1), changing to (2) would be easy, changing to (3) even easier.

**Comment by littledan** [comment]
I don't think this predates Object.freeze chronologically. Maybe @allenwb would know more about the history and motivation.

**Comment by anba** [comment]
Making it frozen was seen as too restrictive -> https://mail.mozilla.org/pipermail/es-discuss/2012-June/023717.html

**Comment by jakobkummerow** [comment]
Thanks for the historical context! My reading of that thread is that there were no strong reasons for any of the options (write-protect individual properties, freeze the array, return a plain mutable array), and the decision for the first option was made to exactly preserve previous behavior and thereby the option to go back to that -- but given that we haven't gone back in 6 years, that argument probably doesn't have much weight any more.

As an aside, that thread also mentions performance. While I don't think that performance matters here, for the record, currently plain arrays are the fastest choice, at least in V8.
While we could change that, and maybe should indeed make sure that freezing an array does not have a performance penalty, I would argue that implementations are generally likely to have particularly efficient (fast and/or memory saving) support for the default case (where no property has any non-standard attributes), whereas storing per-element attributes tends to have a cost: memory for storing and CPU for checking. Again, I'm not saying that's an argument here; I'm just pointing out that "frozen is faster" is also not an argument.

I would still prefer the simplest solution, which is option (3) above.

**Comment by littledan** [comment]
Yeah, given everything, the proposed change 3 sounds good to me.

**Comment by littledan** [comment]
We agreed in the October ECMA-402 meeting to make the change in #278. Now we're just waiting for implementations and tests to merge the change.
